server/src/test/java/org/elasticsearch/repositories/RepositoriesServiceTests.java

```diff
@@ -162,6 +165,61 @@ public void testRepositoriesStatsCanHaveTheSameNameAndDifferentTypeOverTime() {
         assertThat(repositoryStatsTypeB.getRepositoryStats(), equalTo(MeteredRepositoryTypeB.STATS));
     }
 
+    // this can happen when the repository plugin is removed, but repository is still exist
+    public void testHandlesUnknownRepositoryTypeWhenApplyingClusterState() {
+        var repoName = randomAlphaOfLengthBetween(10, 25);
+
+        var clusterState = createClusterStateWithRepo(repoName, "unknown");
+        repositoriesService.applyClusterState(new ClusterChangedEvent("starting", clusterState, emptyState()));
+
+        var repo = repositoriesService.repository(repoName);
+        assertThat(repo, isA(UnknownTypeRepository.class));
+    }
+
+    public void testRemoveUnknownRepositoryTypeWhenApplyingClusterState() {
+        var repoName = randomAlphaOfLengthBetween(10, 25);
+
+        var clusterState = createClusterStateWithRepo(repoName, "unknown");
+        repositoriesService.applyClusterState(new ClusterChangedEvent("starting", clusterState, emptyState()));
+        repositoriesService.applyClusterState(new ClusterChangedEvent("removing repo", emptyState(), clusterState));
+
+        assertThatThrows(
+            () -> repositoriesService.repository(repoName),
+            RepositoryMissingException.class,
+            equalTo("[" + repoName + "] missing")
+        );
+    }
+
+    public void testRegisterRepositoryFailsForUnknownType() {
+        var repoName = randomAlphaOfLengthBetween(10, 25);
+        var request = new PutRepositoryRequest().name(repoName).type("unknown");
+
+        repositoriesService.registerRepository(request, new ActionListener<>() {
+            @Override
+            public void onResponse(AcknowledgedResponse acknowledgedResponse) {
+                fail("Should not register unknown repository type");
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                assertThat(e, RepositoryException.class, equalTo("[" + repoName + "] repository type [unknown] does not exist"));
+            }
+        });
+    }
+
+    private static void assertThatThrows(ThrowingRunnable code, Class<? extends Exception> exceptionType, Matcher<String> messageMatcher) {
+        try {
+            code.run();
+        } catch (Throwable e) {
+            assertThat(e, exceptionType, messageMatcher);
+        }
+    }
+
+    private static void assertThat(Throwable exception, Class<? extends Exception> exceptionType, Matcher<String> messageMatcher) {
+        assertThat(exception, isA(exceptionType));
+        assertThat(exception.getMessage(), messageMatcher);
+    }
```

Comment:
Is there a good common place to move this to? May be ESTestCase?
I am missing a simple way to assert exceptions in hamcrest similarly to what assertj offers: https://assertj.github.io/doc/#assertj-core-exception-assertions

Reply:
TBH we mostly don't care about the exception type, the message is the important bit. I think it'd fit somewhere in `:test:framework:org.elasticsearch.test.hamcrest` if you wanted to add a utility method for this tho.